File: C++/readme.cpp
# 💻 C++ Fundamentals

This repository contains basic concepts and examples of **C++ Programming Fundamentals**.  
It is useful for beginners, students, competitive programmers, and interview preparation.

---

# 📚 Topics Covered

- Introduction to C++
- Variables and Data Types
- Input and Output
- Operators
- Conditional Statements
- Loops
- Functions
- Arrays
- Strings
- Pointers
- Structures
- Object-Oriented Programming
- File Handling
- STL Basics

---

# 🗂 Folder Structure

CPlusPlus-Fundamentals/
│
├── Variables-and-DataTypes/
├── Operators/
├── Loops/
├── Functions/
├── Arrays/
├── Pointers/
├── OOP/
├── File-Handling/
└── README.md

---

# 🚀 Basic C++ Program

## 1. Hello World

```cpp
#include <iostream>
using namespace std;

int main() {
    cout << "Hello, World!" << endl;
    return 0;
}
```

---

# 🔹 Variables and Data Types

```cpp
#include <iostream>
using namespace std;

int main() {
    int age = 22;
    float cgpa = 3.85;
    char grade = 'A';

    cout << "Age: " << age << endl;
    cout << "CGPA: " << cgpa << endl;
    cout << "Grade: " << grade << endl;

    return 0;
}
```

---

# 🔹 Conditional Statement

```cpp
#include <iostream>
using namespace std;

int main() {
    int number;

    cout << "Enter a number: ";
    cin >> number;

    if (number % 2 == 0) {
        cout << "Even Number";
    } else {
        cout << "Odd Number";
    }

    return 0;
}
```

---

# 🔹 Loops Example

```cpp
#include <iostream>
using namespace std;

int main() {

    for (int i = 1; i <= 5; i++) {
        cout << i << endl;
    }

    return 0;
}
```

---

# 🔹 Function Example

```cpp
#include <iostream>
using namespace std;

int add(int a, int b) {
    return a + b;
}

int main() {

    int result = add(5, 3);

    cout << "Sum: " << result << endl;

    return 0;
}
```

---

# 🔹 Array Example

```cpp
#include <iostream>
using namespace std;

int main() {

    int numbers[5] = {10, 20, 30, 40, 50};

    for (int i = 0; i < 5; i++) {
        cout << numbers[i] << " ";
    }

    return 0;
}
```

---

# 🔹 Pointer Example

```cpp
#include <iostream>
using namespace std;

int main() {

    int number = 100;
    int *ptr = &number;

    cout << "Value: " << number << endl;
    cout << "Address: " << ptr << endl;

    return 0;
}
```

---

# 🔹 OOP Example

```cpp
#include <iostream>
using namespace std;

class Student {
public:
    string name;
    int age;

    void display() {
        cout << "Name: " << name << endl;
        cout << "Age: " << age << endl;
    }
};

int main() {

    Student s1;

    s1.name = "Rahim";
    s1.age = 22;

    s1.display();

    return 0;
}
```

---

# 🔹 File Handling Example

```cpp
#include <iostream>
#include <fstream>
using namespace std;

int main() {

    ofstream file("data.txt");

    file << "C++ File Handling Example";

    file.close();

    cout << "File Created Successfully";

    return 0;
}
```

---

# 🎯 Advantages of C++

- Fast and Efficient
- Supports OOP
- Portable Language
- Rich Standard Library
- Used in Game Development
- Useful for Competitive Programming

---

# 💡 Mini Project Ideas

- Calculator
- Student Management System
- Library Management System
- Banking System
- Inventory System

---

# 📖 Learning Goals

- Understand C++ syntax
- Build problem-solving skills
- Learn Object-Oriented Programming
- Prepare for coding interviews
- Improve competitive programming skills

---

# 🤝 Contribution

Contributions are welcome.  
Feel free to fork this repository and improve it.

---

# ⭐ Support

If you found this repository useful, give it a ⭐ on GitHub.

File: Python/Readme.py
# 🐍 Python OOP Fundamentals

This repository contains basic concepts and examples of **Object-Oriented Programming (OOP)** in Python.  
It is designed for beginners, students, and interview preparation.

---

# 📚 Topics Covered

- Class and Object
- Constructor (`__init__`)
- Instance Variables
- Methods
- Inheritance
- Polymorphism
- Encapsulation
- Abstraction
- Method Overriding
- Operator Overloading
- Static Methods
- Class Methods

---

# 🗂 Folder Structure

Python-OOP-Fundamentals/
│
├── Classes-and-Objects/
├── Inheritance/
├── Polymorphism/
├── Encapsulation/
├── Abstraction/
├── Projects/
└── README.md

---

# 🚀 OOP Basics Example

## 1. Class and Object

```python
class Student:
    def __init__(self, name, department):
        self.name = name
        self.department = department

    def show_info(self):
        print(f"Name: {self.name}")
        print(f"Department: {self.department}")


student1 = Student("Rahim", "CSE")
student1.show_info()
```

---

# 🔹 Constructor Example

```python
class Car:
    def __init__(self, brand, model):
        self.brand = brand
        self.model = model

    def display(self):
        print(f"{self.brand} - {self.model}")


car1 = Car("Toyota", "Corolla")
car1.display()
```

---

# 🔹 Inheritance Example

```python
class Animal:
    def sound(self):
        print("Animals make sound")


class Dog(Animal):
    def bark(self):
        print("Dog barks")


dog1 = Dog()
dog1.sound()
dog1.bark()
```

---

# 🔹 Polymorphism Example

```python
class Bird:
    def sound(self):
        print("Birds chirp")


class Cat:
    def sound(self):
        print("Cats meow")


for animal in (Bird(), Cat()):
    animal.sound()
```

---

# 🔹 Encapsulation Example

```python
class BankAccount:
    def __init__(self, balance):
        self.__balance = balance

    def deposit(self, amount):
        self.__balance += amount

    def show_balance(self):
        print("Balance:", self.__balance)


account = BankAccount(1000)
account.deposit(500)
account.show_balance()
```

---

# 🔹 Abstraction Example

```python
from abc import ABC, abstractmethod

class Shape(ABC):

    @abstractmethod
    def area(self):
        pass


class Rectangle(Shape):
    def __init__(self, width, height):
        self.width = width
        self.height = height

    def area(self):
        return self.width * self.height


rect = Rectangle(5, 10)
print("Area:", rect.area())
```

---

# 🔹 Static Method Example

```python
class MathUtils:

    @staticmethod
    def add(a, b):
        return a + b


print(MathUtils.add(5, 3))
```

---

# 🔹 Class Method Example

```python
class University:
    university_name = "Daffodil International University"

    @classmethod
    def show_university(cls):
        print(cls.university_name)


University.show_university()
```

---

# 🎯 Advantages of OOP

- Code Reusability
- Better Code Organization
- Easy Maintenance
- Improved Security
- Modular Programming
- Real-world Modeling

---

# 💡 Mini Project Ideas

- Student Management System
- Library Management System
- Bank Account System
- Employee Management System
- Inventory Management System

---

# 📖 Learning Goals

- Understand OOP concepts clearly
- Build reusable Python code
- Improve problem-solving skills
- Prepare for coding interviews

---

# 🤝 Contribution

Contributions are welcome.  
Feel free to fork this repository and improve it.

---

# ⭐ Support

If you found this repository useful, give it a ⭐ on GitHub.
